src/experimental/information_estimates.cpp

```diff
@@ -0,0 +1,401 @@
+                }
+
+                // Measurement outcomes vector {Pfree, Pocc, Pun}
+                std::vector<kt_double> probabilities {
+                    calculateScanMassProbabilityBetween(distances[1], m_max_sensor_range),
+                    calculateScanMassProbabilityBetween(distances[0], distances[1]),
+                    calculateScanMassProbabilityBetween(0.0f, distances[0])
+                };
+
+                // Appending new measurement outcomes for the current cell
+                appendCellProbabilities(probabilities, cell);
+
+                // Get all the measurement outcomes for the current cell
+                std::vector<std::vector<kt_double>> cell_prob = retrieveCellProbabilities(cell);
+
+                // Compute all the possible combinations for the current cell - algorithm 1
+                std::unordered_map<map_tuple, kt_double, utils::tuple_hash::HashTuple> meas_out_prob = computeMeasurementOutcomesHistogram(cell_prob);
+
+                kt_double cell_mutual_inf = 0.0f;
+                for (auto& pair : meas_out_prob)
+                {
+                    cell_mutual_inf +=  pair.second * measurementOutcomeEntropy(pair.first);
+                }
+
+                // Mutual information of cell x, y given a set of measurements
+                updateCellMutualInformation(1.0 - cell_mutual_inf, cell);
+            }
+        }
+        ++curr_idx;
+    }
+    return m_mutual_grid.sum();
+}
+
+
+void InformationEstimates::appendCellProbabilities(std::vector<kt_double>& measurements, karto::Vector2<int> const & cell)
+{
+    /**
+     * Append measured porbabilities for the given cell
+     * Arguments:
+        * measurements [std::vector<kt_double>]: Vector of LocalizedRangeScan pointers
+        * cell [karto::Vector2<int>]: Cell for appending the data
+     * Return:
+        * Void
+    */
+    std::map<std::vector<int>, std::vector<std::vector<kt_double>>>::iterator it_cell;
+
+    it_cell = m_cell_probabilities.find({cell.GetX(), cell.GetY()});
+    if (it_cell == m_cell_probabilities.end())
+    {
+        // Cell is not present in the map, so append it
+        m_cell_probabilities.insert(std::pair<std::vector<int>, std::vector<std::vector<kt_double>>>(
+            {cell.GetX(), cell.GetY()}, {{measurements[0], measurements[1], measurements[2]}}));
+        m_visited_grid(cell.GetX(), cell.GetY()) = 1;
+    }
+    else
+    {
+        if(m_visited_grid(cell.GetX(), cell.GetY()) == 1)
+        {
+            // Compare the unknown probability, the smallest it is the most information we will have
+            // from the occupied or free state
+            int idx = it_cell->second.size() - 1;
+            if(measurements[2] < it_cell->second[idx][2])
+            {
+                // Replacing
+                it_cell->second[idx][0] = measurements[0];
+                it_cell->second[idx][1] = measurements[1];
+                it_cell->second[idx][2] = measurements[2];
+            }
```

Comment:
@kmilo7204 I lost you here. Why are we comparing mass probabilities from different scans??

Reply:
I think you are right about this one.